### PR TITLE
Feature image based lighting

### PR DESCRIPTION
**Pull request by LaisoEmilio**
Diffuse irradiance part of IBL is implemented, not in shader yet but I created the pull request already to check the file changed that Viktor needs for the dds cubemap fix

### CODE REVIEW

File: resources/shaders/lighting.hlsl [hunk 1]
```diff
@@ -48,10 +48,20 @@ float3 shade_pixel(float3 pos, float3 V, float3 albedo, float metallic, float ro
...
+	hardcoded_light.pos = float3(0.0f, 0.0f, 0.0f);
+	hardcoded_light.rad = 0.0f;
+	hardcoded_light.col = float3(10.0f, 10.0f, 10.0f);
+	hardcoded_light.tid = 1;
+	hardcoded_light.dir = float3(0.0f, -1.0f, 0.0f);
+	hardcoded_light.ang = 0.0f;
+
+	res += shade_light(pos, V, albedo, normal, metallic, roughness, hardcoded_light);
+
+	//for (uint i = 0; i < light_count; i++)
```
Comment (VZout) [inline resources/shaders/lighting.hlsl]: Don't forget to revert this

File: src/d3d12/d3d12_functions.hpp [hunk 2]
```diff
@@ -99,8 +100,12 @@ namespace wr::d3d12
 	void SetName(TextureResource* tex, std::wstring name);
 	void CreateSRVFromTexture(TextureResource* tex);
 	void CreateSRVFromTexture(TextureResource* tex, DescHeapCPUHandle& handle);
+	void CreateUAVFromTexture(TextureResource* tex, DescHeapCPUHandle& handle, unsigned int mip_slice);
+	void CreateRTVFromTexture2D(TextureResource* tex);
+	void CreateRTVFromCubemap(TextureResource* tex);
 	//void CreateUAVFromTexture(TextureResource* tex, DescHeapCPUHandle& handle, unsigned int mip_slice = 0, unsigned int array_slice = 0);
 	void SetShaderTexture(wr::d3d12::CommandList* cmd_list, uint32_t rootParameterIndex, uint32_t descriptorOffset, TextureResource* tex);
```
Comment (VZout) [inline src/d3d12/d3d12_functions.hpp]: SetShaderTexture vs SetShaderUAV. Wanna rename this to SetShaderSRV possibly?

File: src/d3d12/d3d12_resource_pool_texture.cpp [hunk 3]
```diff
@@ -68,23 +87,58 @@ namespace wr
 
 				unstaged_textures.push_back(texture);
 
-				size_t row_pitch, slice_pitch;
+				if (texture->m_width == 1024)
```
Comment (VZout) [inline src/d3d12/d3d12_resource_pool_texture.cpp]: Don't forget to revert tyhis.

File: src/d3d12/d3d12_resource_pool_texture.cpp [hunk 4]
```diff
@@ -96,14 +150,100 @@ namespace wr
...
+	TextureHandle D3D12TexturePool::CreateCubemap(std::string_view name, uint32_t width, uint32_t height, uint32_t mip_levels, Format format, bool allow_render_dest)
+	{
+		auto device = m_render_system.m_device;
+
+		d3d12::desc::TextureDesc desc;
+
+		uint32_t mip_lvl_final = mip_levels;
+
+		// 0 means generate max number of mip levels
+		if (mip_lvl_final == 0)
```
Comment (VZout) [inline src/d3d12/d3d12_resource_pool_texture.cpp]: Miplevels can't be 0. That would mean no subresources at all.

File: src/d3d12/d3d12_resource_pool_texture.cpp [hunk 5]
```diff
@@ -279,19 +419,19 @@ namespace wr
 
 		auto texture = d3d12::CreateTexture(device, &desc, generate_mips);
 
-		texture->m_allocated_memory = static_cast<uint8_t*>(malloc(image.GetPixelsSize()));
+		texture->m_allocated_memory = static_cast<uint8_t*>(malloc(texture->m_needed_memory));
```
Comment (VZout) [inline src/d3d12/d3d12_resource_pool_texture.cpp]: m_allocated_memory maybe shouldn't be uint8t if you need to cast anyway.

File: src/d3d12/d3d12_texture_resources.hpp [hunk 6]
```diff
@@ -22,12 +22,18 @@ namespace wr::d3d12
...
 		ResourceState m_current_state;
-		DescriptorAllocation m_desc_allocation;
+		DescriptorAllocation m_srv_allocation;
+		DescriptorAllocation m_uav_allocation;
+
+		//This allocation can be either 1 or 6 continous descriptors based on m_is_cubemap
+		std::optional<DescriptorAllocation> m_rtv_allocation = std::nullopt;
 
 		uint8_t* m_allocated_memory;
+		size_t m_needed_memory;
```
Comment (VZout) [inline src/d3d12/d3d12_texture_resources.hpp]: please use std::size_t. size_t are C headers and if we all use std:: everywhere its more consistent.

File: src/d3d12/d3d12_textures.cpp [hunk 7]
```diff
@@ -146,9 +152,149 @@ namespace wr::d3d12
...
+		if (tex->m_is_cubemap)
+		{
+			dimension = D3D12_UAV_DIMENSION_TEXTURE2DARRAY;
+
+			uav_desc.Texture2DArray.MipSlice = mip_slice;
+			uav_desc.Texture2DArray.ArraySize = tex->m_array_size;
+		}
+		else
+		{
+			if (tex->m_depth > 1)
```
Comment (VZout) [inline src/d3d12/d3d12_textures.cpp]: Can't you replace this with else if.

File: src/render_tasks/d3d12_cubemap_convolution.hpp [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+
+			data.in_radiance = in_radiance;
+			data.out_irradiance = out_irradiance;
+
+			data.camera_cb_pool = rs.CreateConstantBufferPool(2);
+			data.cb_handle = static_cast<D3D12ConstantBufferHandle*>(data.camera_cb_pool->Create(sizeof(ProjectionView_CBuffer)));
+
+			data.proj_mat = DirectX::XMMatrixPerspectiveFovRH(DirectX::XMConvertToRadians(90.0f), 1.0f, 0.1f, 10.0f);
+
+			data.view_mat[0] = DirectX::XMMatrixLookAtRH(DirectX::XMVectorSet(0.0f, 0.0f, 0.0f, 1.0f),
```
Comment (VZout) [inline src/render_tasks/d3d12_cubemap_convolution.hpp]: You can replace this with a loop. And only define the different forward vectors hardcoded.

File: src/render_tasks/d3d12_cubemap_convolution.hpp [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+								0, static_cast<D3D12ModelPool*>(cube_model->m_model_pool)->GetVertexStagingBuffer()->m_size,
+								n_mesh->m_vertex_staging_buffer_stride);
+
+							d3d12::BindIndexBuffer(cmd_list, static_cast<D3D12ModelPool*>(cube_model->m_model_pool)->GetIndexStagingBuffer(),
+								0, static_cast<D3D12ModelPool*>(cube_model->m_model_pool)->GetIndexStagingBuffer()->m_size);
+
+							d3d12::SetShaderTexture(cmd_list, 2, 0, radiance);
+							d3d12::BindDescriptorHeaps(cmd_list, frame_idx);
+
+							if (n_mesh->m_index_count != 0)
```
Comment (VZout) [inline src/render_tasks/d3d12_cubemap_convolution.hpp]: You can probably guarantee that it takes a indexed cube.

File: src/render_tasks/d3d12_equirect_to_cubemap.hpp [hunk 10]
```diff
@@ -17,69 +18,165 @@
...
+
+			data.camera_cb_pool = rs.CreateConstantBufferPool(2);
+			data.cb_handle = static_cast<D3D12ConstantBufferHandle*>(data.camera_cb_pool->Create(sizeof(ProjectionView_CB)));
+
+			data.in_equirect = in_equirect;
+			data.out_cubemap = out_cubemap;
+
+			data.proj_mat = DirectX::XMMatrixPerspectiveFovRH(DirectX::XMConvertToRadians(90.0f), 1.0f, 0.1f, 10.0f);
+
+			data.view_mat[0] = DirectX::XMMatrixLookAtRH(DirectX::XMVectorSet(0.0f, 0.0f, 0.0f, 1.0f),
```
Comment (VZout) [inline src/render_tasks/d3d12_equirect_to_cubemap.hpp]: ditto, you can do this with a for loop.

File: src/render_tasks/d3d12_equirect_to_cubemap.hpp [hunk 11]
```diff
@@ -17,69 +18,165 @@
...
 
-				d3d12::BindConstantBuffer(cmd_list, d3d12_cb_handle->m_native, 0, frame_idx);
+						d3d12::Bind32BitConstants(cmd_list, &i, 1, 0, 0);
 
-				scene_graph.Render(cmd_list, scene_graph.GetActiveCamera().get());
+						//bind cube and render
+						Model* cube_model = rs.GetSimpleShape(RenderSystem::SimpleShapes::CUBE);
+
+						//Render meshes
+						for (auto& mesh : cube_model->m_meshes)
```
Comment (VZout) [inline src/render_tasks/d3d12_equirect_to_cubemap.hpp]: probably just unnessessary code. You can guarantee the cube is a single mesh.

File: src/render_tasks/d3d12_equirect_to_cubemap.hpp [hunk 12]
```diff
@@ -102,8 +199,8 @@ namespace wr
 		};
 		desc.m_name = "Equirect To Cubemap";
 		desc.m_properties = rt_properties;
-		desc.m_type = RenderTaskType::DIRECT;
-		desc.m_allow_multithreading = true;
+		desc.m_type = RenderTaskType::COMPUTE;
+		desc.m_allow_multithreading = false;
```
Comment (VZout) [inline src/render_tasks/d3d12_equirect_to_cubemap.hpp]: But multithreading is badass....

File: resources/shaders/deferred_composition.hlsl [hunk 13]
```diff
@@ -37,24 +39,84 @@ void main_cs(int3 dispatch_thread_id : SV_DispatchThreadID)
...
+
+		float3 R = reflect(-V, normal);
+		float3 F0 = float3(0.04f, 0.04f, 0.04f);
+		F0 = lerp(F0, albedo, metallic);
+
+		float3 res = float3(0.0f, 0.0f, 0.0f);
+		
+		uint light_count = lights[0].tid >> 2;	//Light count is stored in 30 upper-bits of first light
+
+		for (uint i = 0; i < light_count; i++)
```
Comment (VZout) [inline resources/shaders/deferred_composition.hlsl]: Debug code?

File: src/render_tasks/d3d12_equirect_to_cubemap.hpp (hunk 12 above)
Comment (LaisoEmilio) [reply src/render_tasks/d3d12_equirect_to_cubemap.hpp]: solved.

File: src/d3d12/d3d12_functions.hpp (hunk 2 above)
Comment (LaisoEmilio) [reply src/d3d12/d3d12_functions.hpp]: Done 👍

File: resources/shaders/deferred_composition.hlsl (hunk 13 above)
Comment (LaisoEmilio) [reply resources/shaders/deferred_composition.hlsl]: Done 👍

File: demo/resources.hpp [hunk 14]
```diff
@@ -14,16 +15,24 @@ namespace resources
...
+	static wr::MaterialHandle plastic_material;
+	static wr::MaterialHandle metal_material;
 
-	static wr::TextureHandle loaded_skybox;
-	static wr::TextureHandle loaded_skybox2;
+	static wr::MaterialHandle light_material;
+	static wr::MaterialHandle mirror_material;
+	static wr::TextureHandle equirectangular_environment_map;
+	static wr::TextureHandle cubemap_environment_map;
+	static wr::TextureHandle convoluted_environment_map;
```
[the remark this answers is not in the document]
Comment (LaisoEmilio) [reply demo/resources.hpp]: 👍

**Comment by VZout** [comment]
blind merge lets go

**Pull request merged**